docs/lab/05/index.md

```diff
@@ -0,0 +1,224 @@
+    Ready(T),
+}
+trait Future {
+   type Output;
+   fn poll(&mut self) -> Poll<Self::Output>;
+}
+```
+The `Future` has an `Output` type, that represents the type of the result that it will return once it completes. For `wait_for_falling_edge()`, the Output type is `()` (nothing).
+The function `poll` returns a Poll type, which can either be `Pending`, or `Ready<T>` (T will be output in this case).
+Let's break down what all of this means. A `Future` needs to be checked on, every now and then, to see what its status is. This is the job of the **Executor**. The executor must regularly ask the `Future` if it's completed, or if it needs more time before it can give a result. We can say that the `Future` is `poll`ed, and depending on whether it's ready to give a result or not, it gives its status as `Pending` or `Ready`. If it's still pending, it needs more time before it can return a result, so the executor moves on to poll another `Future`. Whenever the `Future` is completed, it returns `Ready` once polled, and the executor returns execution back to the function where the `Future` completed.
+
+```mermaid
+sequenceDiagram
+    autonumber
+    Executor->>+Future: poll()
+    loop until the Future finishes all the requests to the Hardware
+        Future->>+Hardware: execute_next_action()
+        Hardware-->>Future: in_progress()
+        note right of Hardware: performs the action in parallel
+        Future-->>-Executor: Poll::Pending
+        note over Executor: sleeps until an event arrives
+        note right of Hardware: sends an event when job is done (interrupt)
+        Hardware--)-Executor: event
+        Executor->>+Future: poll()
+    end
+    Future->>Hardware: read_value()
+    Hardware-->>Future: value
+    Future-->>-Executor: Poll::Ready(value)
+```
+
+:::note
+An efficient executor will not poll all tasks. Instead, tasks signal the executor that they are ready to make progress by using a `Waker`.
+:::
+
+TODO: real future in Rust
+
+Under the hood, the Rust compiler is actually transforming our asynchronous function into a state-machine. That is why we can only use `await` inside of an `async` function, because it needs to be treated differently than an ordinary function in order to work asynchronously.
+
+:::info
+Asynchronous programming is widely used in web development. In JavaScript, the equivalent of a `Future` would be a `Promise`.
+:::
+
+Read more about how async/await works in Rust [here](https://rust-lang.github.io/async-book/01_getting_started/01_chapter.html).
+
+## Channels
+
+Up to this point, to be able to share peripherals across multiple tasks, we have been using global `Mutex`s or passing them directly as parameters to the tasks. But there are other, more convenient ways to send data to and from tasks. Instead of having to make global, static variables that are shared by tasks, we could choose to only send the information that we need from one task to another. To achieve this, we can use *channels*.
```

Comment:
```suggestion
Up to this point, to be able to share peripherals or data across multiple tasks, we have been using global `Mutex`s or passing them directly as parameters to the tasks. But there are other, more convenient ways to send data to and from tasks. Instead of having to make global, static variables that are shared by tasks, we could choose to only send the information that we need from one task to another. To achieve this, we can use *channels*.
```